### PR TITLE
Tfchain graphql

### PR DESCRIPTION
**Pull request by AlaaElattar**

### CODE REVIEW

**Review by AhmedHanafy725 — COMMENTED** [review]
- Please make sure only graphql changes are in this PR.
- It has the same major comments as the gridproxy PR.

File: packages/graphql_client/lib/src/contracts.dart [hunk 1]
```diff
@@ -0,0 +1,225 @@
+import 'package:graphql_client/graphql_client.dart';
+import 'package:graphql_client/models.dart';
+
+class TFContracts {
+  final GraphQLClient gqlClient;
+
+  TFContracts(this.gqlClient);
+
+  Future<List<NameContract>> listNameContractsByTwinId(
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/contracts.dart:9]: why list the contracts only by twinId? I think we have some query options to be supported too

Comment (AhmedHanafy725) [reply packages/graphql_client/lib/src/contracts.dart:9]: please check that all methods have query options (including limit, offset, and order)

File: packages/graphql_client/lib/src/node.dart
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/node.dart:1]: let's rename this file to nodes to be consistent with the other files

File: packages/graphql_client/lib/src/farms.dart [hunk 2]
```diff
@@ -0,0 +1,141 @@
+import 'package:graphql_client/graphql_client.dart';
+import '../models.dart';
+
+class TFFarms {
+  final GraphQLClient gqlClient;
+
+  TFFarms(this.gqlClient);
+
+  Future<List<FarmInfo>> listFarms(FarmsQueryOptions? queryOptions,
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/farms.dart:9]:
```suggestion
  Future<List<FarmInfo>> list(FarmsQueryOptions? queryOptions,
```

File: packages/graphql_client/lib/src/farms.dart [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        name: farmsData['name'] ?? '',
+        pricingPolicyID: farmsData['pricingPolicyID'] ?? 0,
+        stellarAddress: farmsData['stellarAddress'] ?? '',
+        twinID: farmsData['twinID'] ?? 0,
+        publicIPs: farmsData['publicIPs'] != null
+            ? (farmsData['publicIPs'] as List<dynamic>)
+                .map((item) => PublicIpsInfo.fromJson(item))
+                .toList()
+            : null,
+      );
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/farms.dart:70]: this should be part of the `FarmInfo`  implementation

File: packages/graphql_client/lib/src/farms.dart [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    }).toList();
+    return farms;
+  }
+
+  //sorts by id_ASC by default
+  //return total count by default
+  Future<FarmsConnectionInfo> getFarmsConnection(
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/farms.dart:77]:
```suggestion
  Future<FarmsConnectionInfo> Connection(
```

File: packages/graphql_client/lib/src/node.dart [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+          interfaces {
+            ips
+            id
+            mac
+            name
+          }
+        }
+''';
+
+  Future<List<Node>> nodes(NodesReturnOptions? returnOptions,
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/node.dart:101]:
```suggestion
  Future<List<Node>> list(NodesReturnOptions? returnOptions,
```

File: packages/graphql_client/lib/src/node.dart [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+          'Invalid data format: Expected a list in "nodes" field, got ${response['data']['nodes'].runtimeType}');
+    }
+    List<Node> nodes =
+        (response['data']['nodes'] as List<dynamic>).map((nodesData) {
+      return Node.fromJson(nodesData as Map<String, dynamic>);
+    }).toList();
+    return nodes;
+  }
+
+  Future<NodeConnectionsInfo> nodesConnection(
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/node.dart:136]:
```suggestion
  Future<NodeConnectionsInfo> Connection(
```

File: packages/graphql_client/lib/src/twins.dart [hunk 7]
```diff
@@ -0,0 +1,102 @@
+import '../graphql_client.dart';
+import '../models.dart';
+
+class TFTwins {
+  final GraphQLClient gqlClient;
+
+  TFTwins(this.gqlClient);
+
+  Future<List<TwinInfo>> twins(
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/src/twins.dart:9]:
```suggestion
  Future<List<TwinInfo>> list(
```

File: packages/graphql_client/lib/models/contracts.dart [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    String returnOptions = '';
+    returnOptions = _addToReturnList(returnOptions, 'id', id);
+    returnOptions = _addToReturnList(returnOptions, 'gridVersion', gridVersion);
+    returnOptions = _addToReturnList(returnOptions, 'contractID', contractID);
+    returnOptions = _addToReturnList(returnOptions, 'twinID', twinID);
+    returnOptions = _addToReturnList(returnOptions, 'state', state);
+    returnOptions = _addToReturnList(returnOptions, 'createdAt', createdAt);
+    returnOptions = _addToReturnList(
+        returnOptions, 'solutionProviderID', solutionProviderID);
+    return returnOptions;
```
Comment (AhmedHanafy725) [inline packages/graphql_client/lib/models/contracts.dart:73]: not sure if we can do that with the reflector but it would be nice if we can

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

**Review by zaelgohary — COMMENTED** [review]
I tried running same example in the docs and I got this error.

image

File: packages/graphql_client/lib/models/node.dart [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+      pricingPolicyID: json['pricingPolicyID'] ?? 0,
+      stellarAddress: json['stellarAddress'] ?? '',
+      twinID: json['twinID'] ?? 0,
+      publicIPs: json['publicIPs'] != null
+          ? (json['publicIPs'] as List<dynamic>)
+              .map((item) => PublicIpsInfo.fromJson(item))
+              .toList()
+          : null,
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/node.dart:51]:
```suggestion
  factory NodeInfo.fromJson(Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/node.dart [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        _addToReturnList(returnOptions, "stellarAddress", stellarAddress);
+    returnOptions = _addToReturnList(returnOptions, "twinID", twinID);
+    returnOptions =
+        _addToReturnList(returnOptions, "publicIPs {", publicIPs != null);
+    returnOptions = _addToReturnList(
+        returnOptions, publicIPs.toString(), publicIPs != null);
+    returnOptions = _addToReturnList(returnOptions, "}", publicIPs != null);
+
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/node.dart:103]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 11]
```diff
@@ -0,0 +1,1720 @@
...
+        _addToReturnList(returnOptions, "serialNumber", serialNumber);
+    returnOptions = _addToReturnList(returnOptions, "createdAt", createdAt);
+    returnOptions = _addToReturnList(returnOptions, "updatedAt", updatedAt);
+    returnOptions =
+        _addToReturnList(returnOptions, "connectionPrice", connectionPrice);
+    if (power != null) returnOptions += power.toString();
+    returnOptions = _addToReturnList(returnOptions, "dedicated", dedicated);
+    returnOptions = _addToReturnList(returnOptions, "extraFee", extraFee);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:994]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 12]
```diff
@@ -0,0 +1,1720 @@
...
+          ? (json['edges'] as List)
+              .map((i) => NodeEdgeInfo.fromJson(i as Map<String, dynamic>))
+              .toList()
+          : null,
+      pageInfo: json['pageInfo'] != null
+          ? PageInfo.fromJson(json['pageInfo'] as Map<String, dynamic>)
+          : null,
+      totalCount: json['totalCount'] ?? 0,
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1146]:
```suggestion
  factory NodeConnectionsInfo.fromJson(Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 13]
```diff
@@ -0,0 +1,1720 @@
...
+  });
+
+  @override
+  String toString() {
+    String returnOptions = "";
+    if (edges != null) returnOptions += edges.toString();
+    if (pageInfo != null) returnOptions += pageInfo.toString();
+    returnOptions = _addToReturnList(returnOptions, "totalCount", totalCount);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1221]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 14]
```diff
@@ -0,0 +1,1720 @@
...
+  String toString() {
+    String returnOptions = "";
+    returnOptions = _addToReturnList(returnOptions, "cursor", cursor);
+    returnOptions = _addToReturnList(returnOptions, "node {", node != null);
+    returnOptions =
+        _addToReturnList(returnOptions, node.toString(), node != null);
+    returnOptions = _addToReturnList(returnOptions, "} ", node != null);
+    if (returnOptions != "") returnOptions = "edges { \n $returnOptions } \n";
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1199]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 15]
```diff
@@ -0,0 +1,1720 @@
...
+    );
+  }
+
+  @override
+  String toString() {
+    return generateToString(this);
+  }
+}
+
+class NodeConnectionsInfo {
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1123]: I got this error when trying to print an instance of <NodeConnectionsInfo> class.

```
Unhandled exception:
Reflecting on un-marked type "NodeConnectionsInfo"
#0      new _InstanceMirrorImpl (package:reflectable/src/reflectable_builder_based.dart:218:9)
#1      ReflectableImpl.reflect (package:reflectable/src/reflectable_builder_based.dart:2559:12)
#2      generateToString (package:graphql_client/src/query_builder.dart:97:45)
#3      NodeConnectionsInfo.toString (package:graphql_client/models/nodes.dart:1063:12)
#4      _StringBase._interpolate (dart:core-patch/string_patch.dart:866:19)
#5      main (file:///home/zainab/repos/tfgrid-sdk-dart/packages/graphql_client/bin/graphql_client.dart:10:21)
#6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 16]
```diff
@@ -0,0 +1,1720 @@
...
+    returnOptions = _addToReturnList(returnOptions, "cru", cru);
+    returnOptions = _addToReturnList(returnOptions, "hru", hru);
+    returnOptions = _addToReturnList(returnOptions, "mru", mru);
+    returnOptions = _addToReturnList(returnOptions, "sru", sru);
+    returnOptions = _addToReturnList(returnOptions, "node {", node != null);
+    returnOptions =
+        _addToReturnList(returnOptions, node.toString(), node != null);
+    returnOptions = _addToReturnList(returnOptions, "} ", node != null);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1311]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 17]
```diff
@@ -0,0 +1,1720 @@
...
+      id: json['id'] ?? "",
+      node: json['node'] != null
+          ? Node.fromJson(json['node'] as Map<String, dynamic>)
+          : null,
+      hru: BigInt.parse(json['hru'] ?? '0'),
+      sru: BigInt.parse(json['sru'] ?? '0'),
+      cru: BigInt.parse(json['cru'] ?? '0'),
+      mru: BigInt.parse(json['mru'] ?? '0'),
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1343]:
```suggestion
  factory NodeResourcesTotal.fromJson(Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 18]
```diff
@@ -0,0 +1,1720 @@
...
+  });
+
+  factory NodeResourcesTotalEdgeInfo.fromJson(Map<String, dynamic> json) {
+    return NodeResourcesTotalEdgeInfo(
+      cursor: json['cursor'] ?? '',
+      node: json['node'] != null
+          ? NodeResourcesTotal.fromJson(json['node'] as Map<String, dynamic>)
+          : null,
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1611]:
```suggestion
  factory NodeResourcesTotalEdgeInfo.fromJson(Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 19]
```diff
@@ -0,0 +1,1720 @@
...
+              .map((i) => NodeResourcesTotalEdgeInfo.fromJson(
+                  i as Map<String, dynamic>))
+              .toList()
+          : null,
+      pageInfo: json['pageInfo'] != null
+          ? PageInfo.fromJson(json['pageInfo'] as Map<String, dynamic>)
+          : null,
+      totalCount: json['totalCount'] ?? 0,
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1644]:
```suggestion
  factory NodeResourcesTotalsConnectionInfo.fromJson(
      Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 20]
```diff
@@ -0,0 +1,1720 @@
...
+  String toString() {
+    String returnOptions = "";
+    returnOptions = _addToReturnList(returnOptions, "cursor", cursor);
+    returnOptions = _addToReturnList(returnOptions, "node {", node != null);
+    returnOptions =
+        _addToReturnList(returnOptions, node.toString(), node != null);
+    returnOptions = _addToReturnList(returnOptions, "} ", node != null);
+    if (returnOptions != "") returnOptions = "edges { \n $returnOptions } \n";
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1697]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/nodes.dart [hunk 21]
```diff
@@ -0,0 +1,1720 @@
...
+  });
+
+  @override
+  String toString() {
+    String returnOptions = "";
+    if (edges != null) returnOptions += edges.toString();
+    if (pageInfo != null) returnOptions += pageInfo.toString();
+    returnOptions = _addToReturnList(returnOptions, "totalCount", totalCount);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/nodes.dart:1719]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/page_info.dart [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+    returnOptions = _addToReturnList(returnOptions, "endCursor", endCursor);
+    returnOptions = _addToReturnList(returnOptions, "hasNextPage", hasNextPage);
+    returnOptions =
+        _addToReturnList(returnOptions, "hasPreviousPage", hasPreviousPage);
+    returnOptions = _addToReturnList(returnOptions, "startCursor", startCursor);
+    if (returnOptions != "") {
+      returnOptions = "pageInfo { \n $returnOptions } \n";
+    }
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/page_info.dart:53]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/power.dart [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+    this.target = false,
+  });
+  @override
+  String toString() {
+    String returnOptions = "";
+    returnOptions = _addToReturnList(returnOptions, "state", state);
+    returnOptions = _addToReturnList(returnOptions, "target", target);
+    if (returnOptions != "") returnOptions = "power { \n $returnOptions } \n";
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/power.dart:61]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/public_config.dart [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+    returnOptions = _addToReturnList(returnOptions, "ipv6", ipv6);
+    returnOptions = _addToReturnList(returnOptions, "gw4", gw4);
+    returnOptions = _addToReturnList(returnOptions, "gw6", gw6);
+    returnOptions = _addToReturnList(returnOptions, "domain", domain);
+    returnOptions = _addToReturnList(returnOptions, "node {", node != null);
+    returnOptions =
+        _addToReturnList(returnOptions, node.toString(), node != null);
+    returnOptions = _addToReturnList(returnOptions, "} ", node != null);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/public_config.dart:36]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

File: packages/graphql_client/lib/models/public_config.dart [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+      node: json['node'] != null
+          ? Node.fromJson(json['node'] as Map<String, dynamic>)
+          : null,
+      ipv4: json['ipv4'] ?? "",
+      ipv6: json['ipv6'] ?? "",
+      gw4: json['gw4'] ?? "",
+      gw6: json['gw6'] ?? "",
+      domain: json['domain'] ?? "",
+    );
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/public_config.dart:71]:
```suggestion
  factory PublicConfig.fromJson(Map<String, dynamic> json) {
    return fromJson(json);
  }
```

File: packages/graphql_client/lib/models/public_ips.dart [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+  @override
+  String toString() {
+    String returnOptions = "";
+    returnOptions = _addToReturnList(returnOptions, "id", id);
+    returnOptions = _addToReturnList(returnOptions, "contractId", contractId);
+    returnOptions = _addToReturnList(returnOptions, "ip", ip);
+    returnOptions = _addToReturnList(returnOptions, "gateway", gateway);
+    return returnOptions;
+  }
```
Comment (zaelgohary) [inline packages/graphql_client/lib/models/public_ips.dart:49]:
```suggestion
  String toString() {
    return generateToString(this);
  }
```

**Review by zaelgohary — CHANGES_REQUESTED** [review]

File: packages/graphql_client/lib/models/farms.dart [hunk 27]
```diff
@@ -0,0 +1,868 @@
...
+  String? idContains;
+  String? idNotContains;
+  String? idContainsInsensitive;
+  String? idNotContainsInsensitive;
+  String? idStartsWith;
+  String? idNotStartsWith;
+  String? idEndsWith;
+  String? idNotEndsWith;
+  List<String>? idIn;
+  List<String>? idNotIn;
```
[the remark this answers is not in the document]
Comment (zaelgohary) [reply packages/graphql_client/lib/models/farms.dart:77]: Opened a refactor issue [here](https://github.com/codescalers/tfgrid-sdk-dart/issues/95)

**Review by zaelgohary — APPROVED** [review]

File: packages/graphql_client/lib/models/nodes.dart [hunk 28]
```diff
@@ -1040,6 +1040,8 @@ class NodeEdgeInfo {
   }
 }
 
+
+@reflector
```
Comment (AlaaElattar) [inline packages/graphql_client/lib/models/nodes.dart:1043]: reflector package not working with nested stuff in classes.

File: packages/stellar_client/lib/src/client.dart [hunk 29]
```diff
@@ -467,4 +474,9 @@ class Client {
       throw Exception("Couldn't get memo text due to ${e}");
     }
   }
+
+  @override
+  String toString() {
```
Comment (AlaaElattar) [inline packages/stellar_client/lib/src/client.dart]: do we need to show this to the user ?

File: packages/graphql_client/lib/models/nodes.dart (hunk 28 above)
Comment (zaelgohary) [reply packages/graphql_client/lib/models/nodes.dart:1043]: image


works fine.

**Comment by AlaaElattar** [comment]
Screenshot 2024-10-21 at 10 53 21 AM

**Comment by AlaaElattar** [comment]
Screenshot 2024-10-21 at 3 22 37 PM

**Comment by AlaaElattar** [comment]
Screenshot 2024-10-21 at 3 25 06 PM

**Comment by AlaaElattar** [comment]
- All return options at any model makes code crashes.

**Comment by zaelgohary** [comment]
> * All return options at any model makes code crashes.

Works normally when I run same example. Did you build reflectable?

image

**Comment by AlaaElattar** [comment]
- When chooseing any var of the return options, the function crashes.
Screenshot 2024-10-23 at 3 43 42 PM

**Comment by AlaaElattar** [comment]
When i added edges here in the return options it failed
```
  final res = await graphQLClient.farms.Connection(FarmsConnectionQueryOptions(orderBy: FarmsOrderByOptions.farmID_ASC), FarmsConnectionReturnOptions(totalCount: true, edges: EdgesReturnOptions(nodeReturnOptions: NodeReturnOptions(id: true))));

```

**Comment by AlaaElattar** [comment]
> When i added edges here in the return options it failed
> 
> ```
>   final res = await graphQLClient.farms.Connection(FarmsConnectionQueryOptions(orderBy: FarmsOrderByOptions.farmID_ASC), FarmsConnectionReturnOptions(totalCount: true, edges: EdgesReturnOptions(nodeReturnOptions: NodeReturnOptions(id: true))));
> ```

same for this one 
```
  final res = await graphQLClient.nodes.Connection(NodesConnectionsReturnOptions(edges: NodeEdgeReturnOptions(cursor: true)),NodesConnectionsQueryOptions());
```

**Comment by AlaaElattar** [comment]
```
  final res = await graphQLClient.nodes.list(NodesReturnOptions(interfaces: InterfacesReturnOptions(name: true)), NodesQueryOptions(limit: 4));
```

**Comment by AlaaElattar** [comment]
```
  final res = await graphQLClient.nodes.nodeResourcesTotal(NodeResourcesTotalReturnOptions(node: NodesReturnOptions(gridVersion: true)), NodeResourcesTotalQueryOptions(orderby: NodeResourcesTotalOrderByOptions.id_ASC));
```

**Comment by AlaaElattar** [comment]
```
  final res = await graphQLClient.nodes.nodeResourcesTotalConnection(NodeResourcesTotalsConnectionReturnOptions(edges: NodeResourcesTotalEdgeReturnOptions(cursor: true, node: NodeResourcesTotalReturnOptions(node: NodesReturnOptions(nodeID: true)))), NodeResourcesTotalsConnectionQueryOptions());
```

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**